server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -1,156 +1,147 @@
+
+    /**
+     * This is the {@link Mono} that we return from {@link ReactorNettyJaxrsServer.Handler#handle(HttpServerRequest,
+     * HttpServerResponse)}
+     */
+    protected final Sinks.Empty<Void> completionSink;
+
+    /**
+     * Indicates that we've starting sending the response bytes.
+     */
+    private volatile boolean started;
+
+    private final NettyOutbound nettyOutbound;
+
+    /**
+     * Used in {@link NettyOutbound#send(Publisher, Predicate)} to trigger flushing the bytes before the {@link
+     * CompletableFuture} returned from `NettyOutbound#then().toFuture()` completes.
+     */
+    private static final Predicate<ByteBuf> FLUSH_ON_EACH_WRITE = bb -> true;
+
+    private final ReactorNettyHttpResponse parentResponse;
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_CANCEL = new WriterException("Response write aborted due to cancellation");
+
+    static final Throwable RESPONSE_WRITE_ABORTED_ON_DISCARD = new WriterException("Response write aborted due to discarded bytes");
+
+    ChunkOutputStream(
+            final ReactorNettyHttpResponse parentResponse,
+            final HttpServerResponse reactorNettyResponse,
+            final Sinks.Empty<Void> completionSink
+    ) {
+        this.completionSink = Objects.requireNonNull(completionSink);
+        this.parentResponse = Objects.requireNonNull(parentResponse);
+        this.nettyOutbound = Objects.requireNonNull(reactorNettyResponse);
+    }
+
+    @Override
+    public void write(int b) {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(byte[] bs, int off, int len) {
+        try {
+            asyncWrite(bs, off, len).toCompletableFuture().get();
+        } catch (final InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new ReactorNettySendException(ie);
+        } catch (final ExecutionException ee) {
+            throw new ReactorNettySendException(ee);
         }
-
-        byte[] bytes = bs;
-        if (offset != 0 || length != bs.length) {
-            bytes = Arrays.copyOfRange(bs, offset, offset + length);
+    }
+
+    @Override
+    public void flush() {
+        try {
+            asyncFlush().get();
+        } catch (final InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new ReactorNettySendException(ie);
+        } catch (final ExecutionException ee) {
+            throw new ReactorNettySendException(ee);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> asyncFlush() {
+        // Everything flows through asyncWrite and we are flushing on each call, so we
+        // will treat that as a no-op for now, assuming that callers of this would have
+        // chained this onto an asyncWrite.  I hope that doesn't mess up SSE..
+        return CompletableFuture.completedFuture(null);
+    }
+
+    static class ReactorNettySendException extends RuntimeException {
```

Comment:
This should likely be moved up to the package and made public. There's no way for a user to catch this explicit exception when it's thrown.